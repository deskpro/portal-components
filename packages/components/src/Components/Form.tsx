import * as React from 'react';
import { Form as FormikForm } from 'formik';
import classNames from 'classnames';

interface IProps {
  className?: string;
  showHover?: boolean;
  children:   React.ReactNode;
}

const Form = ({
  className = '',
  showHover = true,
  ...props
}: IProps) => (
  <FormikForm
    className={classNames('dp-pc_form', className, { 'show-hover': showHover })}
    {...props}
  />);



export default Form;
