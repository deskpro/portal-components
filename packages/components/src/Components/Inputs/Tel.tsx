import * as React from 'react';
import ReactSelect from 'react-select';
import worldCountries from 'world-countries/dist/countries.json';
import classNames from 'classnames';
import { AsYouType, CountryCode } from 'libphonenumber-js';
import Field, { FieldProps, FieldState } from '../Field';
import { SelectContainer } from '../Choices/DropDown';

interface TelProps extends FieldProps {
  excludeCountries:   string[];
  onlyCountries:      string[];
  preferredCountries: string[];
  withIndicator:      boolean;
  placeholder:        string;
}

interface TelState extends FieldState {
  country: CountryCode;
  indicator: string;
  nationalNumber: string;
}

class Tel extends Field<TelProps, TelState> {
  private countries;

  static defaultProps = {
    ...Field.defaultProps,
    excludeCountries:   [],
    onlyCountries:      [],
    preferredCountries: ['us', 'gb'],
    withIndicator:      false,
    placeholder:        '1234 123456',
  };

  constructor(props) {
    super(props);
    this.type = 'tel';
    if (props.withIndicator) {
      this.className = 'dp-pc_field--tel';
      this.state = {
        country:        null,
        indicator:      '',
        nationalNumber: '',
      };
    }
  }

  handleIndicator = (value) => {
    this.setState({
      country:   value.country,
      indicator: value.value
    });
  };

  handleNationalNumber = (e) => {
    const formatted = new AsYouType(this.state.country).input(`+${this.state.indicator}${e.target.value}`);
    const regex = new RegExp(`^\\+${this.state.indicator}\\s*`);
    const nationalNumber = formatted.replace(regex, '');
    this.setState({
      nationalNumber
    });
  };

  static filterIndicator(value, filter) {
    const nameRegex = new RegExp(filter, 'i');
    return value.data.countryName.match(nameRegex) || value.data.value.match(filter);
  }

  loadCountries = () => {
    const { onlyCountries, excludeCountries } = this.props;
    if (!this.countries) {
      this.countries = [];
      Array.from(worldCountries)
        .filter((country) => {
          if (onlyCountries.length) {
            return onlyCountries.includes(country.cca2);
          }
          if (excludeCountries.length) {
            return !excludeCountries.includes(country.cca2);
          }
          return true;
        })
        .forEach((country) => {
          country.idd.suffixes.forEach((code) => {
            this.countries.push({
              value:   code,
              country: country.cca2,
              label:   (
                <span>(+{code})
                  <span className="react-select__description">
                    {country.flag} {country.name.common}
                  </span>
                </span>),
              countryName: country.name.common
            });
          });
        });
    }
    return this.countries;
  };

  renderField = ({ form }) => {
    const {
      name,
      className,
      placeholder,
    } = this.props;
    if (!this.props.withIndicator) {
      return super.renderField(form);
    }
    return <>
      <ReactSelect
        key="indicator"
        value={
          this.loadCountries().find(indicator =>
            indicator.country === this.state.country
            && indicator.value === this.state.indicator)}
        name={name}
        isClearable={false}
        filterOption={Tel.filterIndicator}
        components={{ SelectContainer }}
        onChange={this.handleIndicator}
        onBlur={() => form.setFieldTouched(name, true)}
        options={this.loadCountries()}
        classNamePrefix="react-select"
        placeholder="(1234)"
      />,
      <input
        key="national_number"
        type="tel"
        value={this.state.nationalNumber}
        onChange={this.handleNationalNumber}
        placeholder={placeholder}
        className={classNames('dp-pc_input', className)}
      />
    </>;
  }
}

export default Tel;
