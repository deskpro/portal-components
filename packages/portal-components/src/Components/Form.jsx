import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => <FormikForm {...props} />;

export default Form;
