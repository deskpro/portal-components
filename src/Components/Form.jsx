import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => {
  return <FormikForm {...props} />
};

export default Form;