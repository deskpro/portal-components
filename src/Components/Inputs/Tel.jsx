import Field from '../Field';

class Tel extends Field {
  constructor(props) {
    super(props);
    this.type = 'tel';
  }
}

export default Tel;
