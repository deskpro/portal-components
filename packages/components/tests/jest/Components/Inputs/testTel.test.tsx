import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { withFormik } from 'formik';
import { describe, expect, it } from '@jest/globals';
import Form from '../../../../src/Components/Form';
import Tel from '../../../../src/Components/Inputs/Tel';

interface FormValues {
  tel?: string;
}

interface MyFormikProps {
  tel?: string;
}

describe('>>> Tel --- Snapshot', () => {
  it('+++capturing Snapshot of Tel', () => {
    const App = () => (
      <Form>
        <Tel name="tel" />
      </Form>
    );
    const FormikApp = withFormik<MyFormikProps, FormValues>({
      mapPropsToValues({
        tel
      }) {
        return {
          tel: tel || '0123456789'
        };
      },
      handleSubmit(values, { setSubmitting }) {
        setSubmitting(false);
      }
    })(App);
    const renderedValue = renderer.create(<FormikApp />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
