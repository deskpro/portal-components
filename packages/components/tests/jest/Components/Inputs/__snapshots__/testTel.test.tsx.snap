// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Tel --- Snapshot +++capturing Snapshot of Tel 1`] = `
<form
  className="dp-pc_form show-hover"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="dp-pc_field"
  >
    <label
      className="dp-pc_label"
      htmlFor="field_1"
    >
      
    </label>
    <input
      className="dp-pc_input"
      excludeCountries={Array []}
      id="field_1"
      name="tel"
      onBlur={[Function]}
      onChange={[Function]}
      onlyCountries={Array []}
      placeholder="1234 123456"
      preferredCountries={
        Array [
          "us",
          "gb",
        ]
      }
      required={false}
      type="tel"
      value="0123456789"
      withIndicator={false}
    />
  </div>
</form>
`;
